module/zfs/bpobj.c

```diff
@@ -206,155 +206,213 @@ bpobj_is_empty(bpobj_t *bpo)
+			}
+			mutex_exit(&bpo->bpo_lock);
+			if (bpi->bpi_parent != NULL) {
+				bpobj_close(bpo);
+				kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+			}
+			kmem_free(bpi, sizeof (bpobj_info_t));
+			continue;
+		}
 
-		offset = i * sizeof (uint64_t);
-		blkoff = P2PHASE(i, epb);
+		kmem_free(bpi, sizeof (bpobj_info_t));
+		ASSERT(bpo->bpo_havecomp);
 
-		if (dbuf == NULL || dbuf->db_offset > offset) {
-			if (dbuf)
-				dmu_buf_rele(dbuf, FTAG);
-			err = dmu_buf_hold(bpo->bpo_os,
-			    bpo->bpo_phys->bpo_subobjs, offset, FTAG, &dbuf, 0);
-			if (err)
-				break;
+		dmu_object_info_t doi;
+		err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs,
+		    &doi);
+		if (err) {
+			mutex_exit(&bpo->bpo_lock);
+			return (err);
 		}
+		ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
+		int epb = doi.doi_data_block_size / sizeof (uint64_t);
+
+		for (i = 0; i < bpo->bpo_phys->bpo_num_subobjs; i++) {
+			uint64_t offset = i * sizeof (uint64_t);
+			uint64_t blkoff = P2PHASE(i, epb);
+
+			if (dbuf == NULL || dbuf->db_offset > offset) {
+				if (dbuf)
+					dmu_buf_rele(dbuf, FTAG);
+				err = dmu_buf_hold(bpo->bpo_os,
+				    bpo->bpo_phys->bpo_subobjs, offset, FTAG,
+				    &dbuf, 0);
+				if (err)
+					break;
+			}
 
-		ASSERT3U(offset, >=, dbuf->db_offset);
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
+			ASSERT3U(offset, >=, dbuf->db_offset);
+			ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
 
-		objarray = dbuf->db_data;
-		err = bpobj_open(&sublist, bpo->bpo_os, objarray[blkoff]);
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
+			uint64_t *objarray = dbuf->db_data;
+			bpobj_t *sublist = kmem_alloc(sizeof (bpobj_t),
+			    KM_SLEEP);
+			err = bpobj_open(sublist, bpo->bpo_os,
+			    objarray[blkoff]);
+			if (err)
 				break;
-			}
+			list_insert_head(&stack, bpi_alloc(sublist, bpo,
+			    objarray[blkoff], i));
 		}
-		err = bpobj_iterate_impl(&sublist, func, arg, tx, free);
-		if (free) {
-			VERIFY3U(0, ==, bpobj_space(&sublist,
-			    &used_after, &comp_after, &uncomp_after));
-			bpo->bpo_phys->bpo_bytes -= used_before - used_after;
-			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
-			bpo->bpo_phys->bpo_comp -= comp_before - comp_after;
-			bpo->bpo_phys->bpo_uncomp -=
-			    uncomp_before - uncomp_after;
+		if (dbuf) {
+			dmu_buf_rele(dbuf, FTAG);
+			dbuf = NULL;
 		}
 
-		bpobj_close(&sublist);
-		if (err)
-			break;
-		if (free) {
-			err = dmu_object_free(bpo->bpo_os,
-			    objarray[blkoff], tx);
-			if (err)
-				break;
-			bpo->bpo_phys->bpo_num_subobjs--;
-			ASSERT3S(bpo->bpo_phys->bpo_num_subobjs, >=, 0);
-		}
-	}
-	if (dbuf) {
-		dmu_buf_rele(dbuf, FTAG);
-		dbuf = NULL;
-	}
-	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os,
-		    bpo->bpo_phys->bpo_subobjs,
-		    (i + 1) * sizeof (uint64_t), DMU_OBJECT_END, tx));
+		mutex_exit(&bpo->bpo_lock);
```

Comment:
in this case, when do we call `bpobj_close(bpo)`?  It looks like we're leaving it open, and it's pointed to by the childrens' `bpi_parent`, but I don't see where we close it.

Reply:
Right, but what about the intermediate bpobj's?  I.e. dp_free_bpobj points to a bpobj which points to a bpobj which points to this one.

Reply:
I think you're referring to the code that's now on line 373.  So bpi_parent always refers to the bpi_bpo of another bpi?  I wonder if that would be more clear if bpi_parent was a `bpobj_info_t *` rather than a `bpobj_t *`